Main.hs

```diff
@@ -350,10 +346,14 @@ compileTerm v =
       hsLambda (absName b) <$> underAbstr_ b compileTerm
     Lam v b | visible v ->
       -- System-inserted lambda, no need to preserve the name.
-      -- TODO: most likely a section, so should really reconstruct that.
       underAbstraction_ b $ \ body -> do
         x <- showTCM (Var 0 [])
-        hsLambda x <$> compileTerm body
+        let hsx = Hs.Var () $ Hs.UnQual () (hsName x)
+        body <- compileTerm body
+        return $ case body of
+          Hs.InfixApp _ a op b
+            | a == hsx -> Hs.RightSection () op b -- System-inserted visible lambdas can only come from sections
```

Comment:
`x` may not be system-inserted from section, so this again imposes sections to the user, e.g.
```agda
test₁ : Nat → Nat
test₁ = λ x → 5 + x
```
compiles to:
```haskell
test₁ :: Integer -> Integer
test₁ = (5 +)
```

Reply:
That's a left section so doesn't happen here. The problem is that Agda eta-contracts the lambda before it gets to us. Right-sections can only come from actual user-written sections.